File: 20-ValidParentheses/20-ValidParentheses.java
// Last updated: 26/03/2026, 16:25:19
import java.util.*;

class Solution {
    public boolean isValid(String s) {

        Stack<Character> stack = new Stack<>();

        for (char c : s.toCharArray()) {

            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else {

                if (stack.isEmpty())
                    return false;

                char top = stack.pop();

                if ((c == ')' && top != '(') ||
                    (c == '}' && top != '{') ||
                    (c == ']' && top != '[')) {
                    return false;
                }
            }
        }

        return stack.isEmpty();
    }
}

File: 20-ValidParentheses/20-ValidParentheses.py
# Last updated: 05/08/2026, 10:33:28
class Solution(object):
    def isValid(self, s):
        st = []

        for i in s:
            if i == "(" or i == "[" or i == "{":
                st.append(i)
            else:

                if not st:
                    return False
                if (i == ")" and st.pop() != "(") or \
                   (i == "]" and st.pop() != "[") or \
                   (i == "}" and st.pop() != "{"):
                    return False

        return len(st) == 0